### PR TITLE
Add consent dialog

### PR DESCRIPTION
**Pull request by neubig**
Co-authored with OpenDevin.

Fixes https://github.com/OpenDevin/OpenDevin/issues/2167
Fixes https://github.com/OpenDevin/OpenDevin/issues/2033

This adds a consent dialog to sharing OpenDevin chat histories. See the attached video for the UI experience.

https://github.com/OpenDevin/OpenDevin/assets/398875/a577c700-f8a2-4e29-bc11-d38b9cc3d1bd

### CODE REVIEW

**Comment by amanape** [comment]
@neubig You may find it useful to extend the [`<BaseModal>`](https://github.com/OpenDevin/OpenDevin/blob/04d7354501a25006fde26543b1da7ab4a5fdacd5/frontend/src/components/modals/base-modal/BaseModal.tsx) component that exists. Some example usages are the [`<SettingsModal>`](https://github.com/OpenDevin/OpenDevin/blob/04d7354501a25006fde26543b1da7ab4a5fdacd5/frontend/src/components/modals/settings/SettingsModal.tsx) and the [`<LoadPreviousSessionModal>`](https://github.com/OpenDevin/OpenDevin/blob/04d7354501a25006fde26543b1da7ab4a5fdacd5/frontend/src/components/modals/load-previous-session/LoadPreviousSessionModal.tsx)

**Comment by neubig** [comment]
Hi @amanape , thanks for the feedback! I think this is ready to review if you have a moment.

File: frontend/src/components/chat/ChatInterface.tsx [hunk 1]
```diff
@@ -53,31 +54,27 @@ function ChatInterface() {
...
-    // https://github.com/OpenDevin/OpenDevin/issues/2033
-    const data: FeedbackData = {
-      email: "NOT_PROVIDED",
-      token: getToken(),
-      feedback,
-      permissions: "private",
+  const feedbackVersion = "1.0";
+  const [feedback, setFeedback] = React.useState<Feedback>({
+    version: feedbackVersion,
+  } as Feedback);
```
Comment (amanape) [inline frontend/src/components/chat/ChatInterface.tsx:60]:
```suggestion
  const [feedback, setFeedback] = React.useState<Feedback>({
    email: "",
    feedback: "positive",
    permissions: "private",
    trajectory: [],
    token: "",
    version: feedbackVersion,
  });
```

File: frontend/src/components/chat/ChatInterface.tsx [hunk 2]
```diff
@@ -164,6 +159,12 @@ function ChatInterface() {
         disabled={curAgentState === AgentState.LOADING}
         onSendMessage={handleSendMessage}
       />
+      <FeedbackModal
+        feedback={feedback}
+        setFeedback={setFeedback}
+        isOpen={feedbackModalIsOpen}
+        onOpenChange={onFeedbackModalOpenChange}
+      />
```
Comment (amanape) [inline frontend/src/components/chat/ChatInterface.tsx:167]:
```suggestion
      <FeedbackModal
        feedback={feedback}
        handleEmailChange={handleEmailChange}
        handlePermissionsChange={handlePermissionsChange}
        isOpen={feedbackModalIsOpen}
        onOpenChange={onFeedbackModalOpenChange}
      />
```

File: frontend/src/components/modals/feedback/FeedbackModal.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  isOpen: boolean;
+  onOpenChange: (isOpen: boolean) => void;
+}
+
+function FeedbackModal({
+  feedback,
+  setFeedback,
+  isOpen,
+  onOpenChange,
+}: FeedbackModalProps) {
```
Comment (amanape) [inline frontend/src/components/modals/feedback/FeedbackModal.tsx:21]:
```suggestion
interface FeedbackModalProps {
  feedback: Feedback;
  handleEmailChange: (key: string) => void;
  handlePermissionsChange: (permissions: "public" | "private") => void;
  isOpen: boolean;
  onOpenChange: (isOpen: boolean) => void;
}

function FeedbackModal({
  feedback,
  handleEmailChange,
  handlePermissionsChange,
  isOpen,
  onOpenChange,
}: FeedbackModalProps) {
```

File: frontend/src/components/modals/feedback/FeedbackModal.tsx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      });
+  };
+
+  const handleEmailChange = (key: string) => {
+    setFeedback({ ...feedback, email: key } as Feedback);
+  };
+
+  const handlePermissionsChange = (permissions: "public" | "private") => {
+    setFeedback({ ...feedback, permissions } as Feedback);
+  };
```
Comment (amanape) [inline frontend/src/components/modals/feedback/FeedbackModal.tsx:50]:
```suggestion

```

File: frontend/src/components/chat/ChatInterface.tsx (hunk 1 above)
Comment (amanape) [reply frontend/src/components/chat/ChatInterface.tsx:60]: It is more type-safe to defined all the defaults, or, set the state type to be `Partial<Feedback>`

**Review by amanape — COMMENTED** [review]
Looks good! Though it can be improved by moving the `setFeedback` logic from the modal to the parent component that defines the `setFeedback` hook:

```ts
  const handleEmailChange = (key: string) => {
    setFeedback((prev) => ({ ...prev, email: key }));
  };

  const handlePermissionsChange = (permissions: "public" | "private") => {
    setFeedback((prev) => ({ ...prev, permissions }));
  };
```

All the suggested changes below are related to this edit

**Review by amanape — APPROVED** [review]

**Comment by neubig** [comment]
Thanks, all fixed!

**Pull request merged**